File: sim/character.py
from typing import Dict, List, Optional, Set
import math

import sim.bardic_inspiration
from util.util import prof_bonus
from sim.core_feats import Vex, Topple, Graze
from sim.events import AttackRollArgs, AttackArgs, AttackResultArgs
from sim.event_loop import EventLoop
from util.log import log
from sim.spells import Spellcasting, Spellcaster
from sim.attack import WeaponAttack, SpellAttack

import sim
import sim.maneuvers
import sim.events
import sim.attack
import sim.spells
import sim.weapons
import sim.feat
import sim.target
import sim.weapons
import sim.resource

STATS = ["str", "dex", "con", "int", "wis", "cha"]
DEFAULT_STAT_MAX = 20


class Character:
    def init(
        self,
        level: int,
        stats: List[int],
        base_feats: Optional[List["sim.feat.Feat"]] = None,
        spellcaster: Spellcaster = Spellcaster.NONE,
        spell_mod: "sim.Stat" = "none",
    ):
        base_feats = base_feats or []
        self.level = level
        self.prof = prof_bonus(level)
        self.stats = {
            "str": stats[0],
            "dex": stats[1],
            "con": stats[2],
            "int": stats[3],
            "wis": stats[4],
            "cha": stats[5],
        }
        self.stat_max = {stat: DEFAULT_STAT_MAX for stat in STATS}
        self.minions: List[Character] = []
        self.events = EventLoop()
        self.effects: Set[str] = set()
        self.spells = Spellcasting(self, spell_mod, [(spellcaster, level)])
        self.masteries: Set["sim.weapons.WeaponMastery"] = set()
        self.used_bonus = False
        self.ki = sim.resource.Resource(self, short_rest=True)
        self.maneuvers = sim.maneuvers.Maneuvers()
        self.sorcery = sim.resource.Resource(self)
        self.bardic_inspiration = sim.bardic_inspiration.BardicInspiration(self)
        self.channel_divinity = sim.resource.Resource(self, short_rest=True)
        self.metamagics: Set[str] = set()
        self.class_levels: Dict[str, int] = dict()
        self.resources: Dict[str, "sim.resource.Resource"] = dict()

        self.feats: List["sim.feat.Feat"] = []
        for feat in [Vex(), Topple(), Graze()]:
            self.add_feat(feat)
        for feat in base_feats:
            self.add_feat(feat)

    def add_feat(self, feat: "sim.feat.Feat"):
        feat.apply(self)
        self.feats.append(feat)
        self.events.add(feat, feat.events())
        print(self.feats)

    def has_feat(self, name: str):
        return name in self.feats

    def stat(self, stat: "sim.Stat"):
        if stat == "none":
            return 10
        return self.stats[stat]

    def mod(self, stat: "sim.Stat"):
        if stat == "none":
            return 0
        return (self.stats[stat] - 10) // 2

    def increase_stat_max(self, stat: "sim.Stat", amount: int):
        if stat == "none":
            return
        self.stat_max[stat] += amount

    def increase_stat(self, stat: "sim.Stat", amount: int):
        if stat == "none":
            return
        new_val = self.stats[stat] + amount
        if new_val > self.stat_max[stat]:
            new_val = self.stat_max[stat]
        self.stats[stat] = new_val

    def dc(self, stat: "sim.Stat"):
        return self.mod(stat) + self.prof + 8

    def add_minion(self, minion):
        self.minions.append(minion)

    def remove_minion(self, minion):
        self.minions.remove(minion)

    def add_effect(self, effect: str):
        self.effects.add(effect)

    def remove_effect(self, effect: str):
        if effect in self.effects:
            self.effects.remove(effect)

    def has_effect(self, effect: str):
        return effect in self.effects

    def has_class_level(self, class_name: str, level: int):
        return (
            class_name in self.class_levels and self.class_levels[class_name] >= level
        )

    def add_class_level(self, class_name: str, level: int):
        self.class_levels[class_name] = level

    def use_bonus(self, source: str):
        if not self.used_bonus:
            log.record(f"Bonus ({source})", 1)
            self.used_bonus = True
            return True
        return False

    def add_resource(self, name: str, short_rest: bool = False):
        self.resources[name] = sim.resource.Resource(self, short_rest=short_rest)

    def has_resource(self, name: str):
        return name in self.resources and self.resources[name].num > 0

    # =============================
    #       LIFECYCLE EVENTS
    # =============================

    def begin_turn(self, target: "sim.target.Target"):
        log.record("Turn", 1)
        self.actions = 1
        self.used_bonus = False
        self.events.emit("begin_turn", target)

    def end_turn(self, target: "sim.target.Target"):
        self.events.emit("end_turn", target)
        if not self.used_bonus:
            log.record(f"Bonus (None)", 1)

    def turn(self, target: "sim.target.Target"):
        self.begin_turn(target)
        self.events.emit("before_action", target)
        while self.actions > 0:
            self.events.emit("action", target)
            self.actions -= 1
        self.events.emit("after_action", target)
        self.end_turn(target)
        for minion in self.minions:
            minion.turn(target)

    def short_rest(self):
        self.effects = set()
        self.events.emit("short_rest")

    def long_rest(self):
        self.short_rest()
        self.events.emit("long_rest")

    def enemy_turn(self, target: "sim.target.Target"):
        self.events.emit("enemy_turn", target)

    # ============================
    #      ATTACKS
    # ============================

    def weapon_attack(
        self,
        target: "sim.target.Target",
        weapon: "sim.weapons.Weapon",
        tags: Optional[List[str]] = None,
    ):
        attack = WeaponAttack(weapon)
        self.attack(target=target, attack=attack, weapon=weapon, tags=tags)

    def spell_attack(
        self,
        target: "sim.target.Target",
        spell: "sim.spells.Spell",
        damage: Optional["sim.attack.DamageRoll"] = None,
        callback: Optional["sim.events.AttackResultCallback"] = None,
        is_ranged: bool = False,
    ):
        attack = SpellAttack(
            spell,
            callback=callback,
            is_ranged=is_ranged,
            damage=damage,
        )
        self.attack(target=target, attack=attack, spell=spell)

    def attack(
        self,
        target: "sim.target.Target",
        attack: "sim.attack.Attack",
        weapon: Optional["sim.weapons.Weapon"] = None,
        spell: Optional["sim.spells.Spell"] = None,
        tags: Optional[List[str]] = None,
    ):
        args = AttackArgs(
            target=target, attack=attack, weapon=weapon, spell=spell, tags=tags
        )
        log.record(f"Attack ({args.attack.name})", 1)
        self.events.emit("before_attack")
        to_hit = args.attack.to_hit(self)
        roll_result = self.attack_roll(attack=args, to_hit=to_hit)
        roll = roll_result.roll()
        min_crit = (
            args.attack.min_crit()
            if roll_result.min_crit is None
            else roll_result.min_crit
        )
        crit = roll >= min_crit
        roll_total = roll + to_hit + roll_result.situational_bonus
        hit = roll_total >= args.target.ac
        result = AttackResultArgs(attack=args, hit=hit, crit=crit, roll=roll)
        if hit:
            log.record(f"Hit ({args.attack.name})", 1)
        else:
            log.record(f"Miss ({args.attack.name})", 1)
        if crit:
            log.record(f"Crit ({args.attack.name})", 1)
        args.attack.attack_result(result, self)
        self.events.emit("attack_result", result)
        for damage in result.damage_rolls:
            if crit:
                damage.dice = 2 * damage.dice
            self.do_damage(
                target=args.target,
                damage=damage,
                attack=args,
                spell=args.spell,
                multiplier=result.dmg_multiplier,
            )

    def attack_roll(self, attack: AttackArgs, to_hit: int) -> AttackRollArgs:
        target = attack.target
        args = AttackRollArgs(attack=attack, to_hit=to_hit)
        if target.stunned:
            args.adv = True
        if target.prone:
            if attack.attack.is_ranged():
                args.disadv = True
            else:
                args.adv = True
        if target.semistunned:
            args.adv = True
            target.semistunned = False
        self.events.emit("attack_roll", args)
        return args

    def do_damage(
        self,
        target: "sim.target.Target",
        damage: "sim.attack.DamageRoll",
        attack: Optional["sim.events.AttackArgs"] = None,
        spell: Optional["sim.spells.Spell"] = None,
        multiplier: float = 1.0,
    ):
        args = sim.events.DamageRollArgs(
            target=target,
            damage=damage,
            attack=attack,
            spell=spell,
        )
        self.events.emit("damage_roll", args)
        target.damage_source(
            damage.source, math.floor(args.damage.total() * multiplier)
        )

File: web/src/ts/sim/Character.ts
import {
    CharacterEvent,
    CharacterEventData,
    CharacterEventMapping,
    CharacterEventName,
} from "./events/CharacterEvent"
import { Feat } from "./Feat"
import { EventLoop } from "../util/EventLoop"
import { Class, Stat, WeaponMastery } from "./types"
import { Resource } from "./Resource"
import { Target } from "./Target"
import { Weapon } from "./Weapon"
import { Attack, WeaponAttack } from "./Attack"
import { DamageRoll } from "./helpers/DamageRoll"
import { Spellcasting } from "./spells/Spellcasting"
import { Spell } from "./spells/Spell"
import { ActionEvent } from "./events/ActionEvent"
import { AttackEvent } from "./events/AttackEvent"
import { AttackResultEvent } from "./events/AttackResultEvent"
import { AttackRollEvent } from "./events/AttackRollEvent"
import { DamageRollEvent } from "./events/DamageRollEvent"

const DEFAULT_STAT_MAX = 20

export class Character {
    // Stats
    level: number = 0
    stats: Record<Stat, number> = {
        str: 10,
        dex: 10,
        con: 10,
        int: 10,
        wis: 10,
        cha: 10,
        none: 10,
    }
    statMax: Record<Stat, number> = {
        str: DEFAULT_STAT_MAX,
        dex: DEFAULT_STAT_MAX,
        con: DEFAULT_STAT_MAX,
        int: DEFAULT_STAT_MAX,
        wis: DEFAULT_STAT_MAX,
        cha: DEFAULT_STAT_MAX,
        none: DEFAULT_STAT_MAX,
    }

    events: EventLoop<CharacterEventName, CharacterEventMapping> =
        new EventLoop()
    effects: Set<string> = new Set()
    masteries: Set<WeaponMastery> = new Set()
    classLevels: Map<Class, number> = new Map()
    feats: Array<Feat> = []
    minions: Set<Character> = new Set()

    spells: Spellcasting = new Spellcasting({
        character: this,
    })

    bonus: Resource = new Resource({
        name: "Bonus",
        character: this,
        initialMax: 1,
    })
    // TODO: Add other class resources

    // TODO: Handle actions better
    actions: number = 1

    addFeat<T extends CharacterEvent>(feat: Feat): void {
        this.feats.push(feat)
        feat.coreApply(this)
    }

    stat(stat: Stat): number {
        return this.stats[stat]
    }

    mod(stat: Stat): number {
        return Math.floor((this.stat(stat) - 10) / 2)
    }

    increaseStatMax(stat: Stat, amount: number): void {
        this.statMax[stat] += amount
    }

    increaseStat(stat: Stat, amount: number): void {
        this.stats[stat] += amount
        if (this.stats[stat] > this.statMax[stat]) {
            this.stats[stat] = this.statMax[stat]
        }
    }

    dc(stat: Stat): number {
        return this.mod(stat) + this.prof() + 8
    }

    prof(): number {
        return Math.floor((this.level - 1) / 4) + 2
    }

    addMinion(minion: Character): void {
        this.minions.add(minion)
    }

    removeMinion(minion: Character): void {
        this.minions.delete(minion)
    }

    hasEffect(effect: string): boolean {
        return this.effects.has(effect)
    }

    addEffect(effect: string): void {
        this.effects.add(effect)
    }

    removeEffect(effect: string): void {
        this.effects.delete(effect)
    }

    getClassLevel(class_: Class): number {
        return this.classLevels.get(class_) ?? 0
    }

    hasClassLevel(class_: Class, level: number): boolean {
        return (this.getClassLevel(class_) ?? 0) >= level
    }

    addClassLevel(class_: Class, level: number): void {
        this.classLevels.set(class_, this.getClassLevel(class_) + level)
    }

    // =============================
    //       LIFECYCLE EVENTS
    // =============================

    turn(target: Target): void {
        this.actions = 1
        this.bonus.reset()
        this.events.emit("begin_turn", { name: "begin_turn", target })
        this.events.emit("before_action", { name: "before_action", target })
        while (this.actions > 0) {
            this.events.emit("action", new ActionEvent({ target }))
            this.actions -= 1
        }
        this.events.emit("after_action", { name: "after_action", target })
        this.events.emit("end_turn", { name: "end_turn", target })
        this.bonus.reset()
        for (const minion of this.minions) {
            minion.turn(target)
        }
    }

    shortRest(): void {
        this.effects = new Set()
        this.events.emit("short_rest", { name: "short_rest" })
    }

    longRest(): void {
        this.shortRest()
        this.events.emit("long_rest", { name: "long_rest" })
    }

    enemyTurn(target: Target): void {
        this.events.emit("enemy_turn", { name: "enemy_turn", target })
    }

    // =============================
    //       WEAPON EVENTS
    // =============================

    weaponAttack(args: {
        target: Target
        weapon: Weapon
        tags?: string[]
    }): void {
        const attack = new WeaponAttack({
            weapon: args.weapon,
            tags: args.tags,
        })
        this.attack({ target: args.target, attack })
    }

    attack(args: { target: Target; attack: Attack }): void {
        const { target, attack } = args
        const attackData = new AttackEvent({ target, attack })
        this.events.emit("before_attack", { name: "before_attack" })
        const toHit = attack.toHit(this)
        const rollResult = this.attackRoll(attackData, toHit)
        const roll = rollResult.roll()
        const minCrit = rollResult.minCrit
            ? rollResult.minCrit
            : attack.minCrit()
        const crit = roll >= minCrit
        const rollTotal = roll + toHit + rollResult.situationalBonus
        const hit = rollTotal >= target.ac
        const attackResult = new AttackResultEvent({
            attack: attackData,
            hit,
            crit,
            roll,
        })
        args.attack.attackResult(attackResult, this)
        this.events.emit("attack_result", attackResult)
        for (const damage of attackResult.damageRolls) {
            if (crit) {
                damage.dice = damage.dice.concat(damage.dice)
            }
            this.doDamage({ target, damage, attack: attackData, multiplier: 1 })
        }
    }

    attackRoll(attackData: AttackEvent, toHit: number): AttackRollEvent {
        const { target, attack } = attackData
        const data = new AttackRollEvent({ attack: attackData, toHit })
        // TODO: Stunned and semistunned
        if (target.prone) {
            if (attack.isRanged()) {
                data.disadv = true
            } else {
                data.adv = true
            }
        }
        this.events.emit("attack_roll", data)
        return data
    }

    doDamage(args: {
        target: Target
        damage: DamageRoll
        attack?: AttackEvent
        multiplier?: number
        spell?: Spell
    }): void {
        const { target, damage, attack, spell } = args
        const multiplier = args.multiplier ?? 1
        const damageData = new DamageRollEvent({
            target,
            damage,
            attack,
            spell,
        })
        this.events.emit("damage_roll", damageData)
        target.addDamage(Math.floor(damage.total() * multiplier))
    }
}
